Port SaveTokensAsClaims to the OpenID Connect middleware

PinpointTownes opened the pull request:
Fixes https://github.com/aspnet/Security/issues/433.

/cc @Eilon @HaoK @Tratcher

PinpointTownes commented on src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectHandler.cs:
```diff
@@ -89,6 +89,12 @@ public OpenIdConnectHandler(HttpClient backchannel)
+                var ticket = await HandleAuthenticateOnceAsync();
+                if (ticket != null)
+                {
+                    message.IdTokenHint = ticket.Principal?.FindFirst(OpenIdConnectParameterNames.IdToken)?.Value;
```
@leastprivilege I know you're massively using the `RedirectToIdentityProvider` event to set the `id_token_hint` in your samples. Any objection if I add it directly to the OIDC handler?

leastprivilege replied:
Doesn't this have a pretty strong assumption how the original id token was preserved.  Isn't the Katana middleware using the bootstrap context for the same purpose?

leastprivilege replied:
maybe I am missing the context

leastprivilege replied:
There are btw situations where you don't want to send the id_token on logout - I had a couple of cases where it was too big for the query string and IE bombed.

PinpointTownes replied:
> Doesn't this have a pretty strong assumption how the original id token was preserved. Isn't the Katana middleware using the bootstrap context for the same purpose?

Yep, but this feature is disabled by default (`TokenValidationParameters.SaveSigninToken` is `false`) and we'd have to parse/validate the id_token a second time since it'd be stored in the BC as a string (https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/KDev/src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs#L639).

PinpointTownes replied:
> There are btw situations where you don't want to send the id_token on logout - I had a couple of cases where it was too big for the query string and IE bombed.

True. Maybe a new option - `FlowIdentityTokenHint` - would help mitigating this issue?

FYI: this middleware now supports POST logout requests via an auto-post form, which solves this kind of issues. But since it's not supported everywhere, I guess the additional option is worth it.

leastprivilege replied:
OK. Sounds good.

As long as the RedirectToIdentityProvider notification is not going away (for both directions).

PinpointTownes replied:
Actually, `RedirectToIdentityProvider` is already gone... replaced by `RedirectToAuthenticationEndpoint` and `RedirectToEndSessionEndpoint` :smile:

leastprivilege replied:
What's the latter?

PinpointTownes replied:
It's like the old `RedirectToIdentityProvider`, but it's only invoked for signout.

leastprivilege replied:
got it.

Tratcher commented on src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectHandler.cs:
```diff
@@ -549,6 +561,25 @@ public OpenIdConnectHandler(HttpClient backchannel)
                     return null;
                 }
                 ticket = authorizationCodeReceivedContext.AuthenticationTicket;
+
```
`message = authenticationValidatedContext.ProtocolMessage;` since you're using it after the event.

PinpointTownes replied:
Okay. Note that this change will be useless whence you implement the hybrid flow, since you'll have to use the token response and not the authorization response at L. 569.

Tratcher commented:
Just the one minor comment. Then :shipit:

PinpointTownes commented:
@Tratcher should I add the option mentioned in https://github.com/aspnet/Security/pull/463#discussion_r39830105 to help mitigating the URL length issues mentioned by @leastprivilege?

Tratcher commented:
How big is the hint?

Note the hint probably won't flow if you're using separate app cookies.

PinpointTownes commented:
> How big is the hint?

The hint corresponds to the identity token returned by the authorization server during the authorization response or the token response, so potentially big.

> Note the hint probably won't flow if you're using separate app cookies.

Instead, should we call the authentication manager to resolve the identity returned by the cookie middleware?

Tratcher commented:
> Instead, should we call the authentication manager to resolve the identity returned by the cookie middleware?
You can try, but the app identity is unlikely to maintain the hint. Now that I think about it, HandleAuthenticateOnceAsync is unlikely to ever provide a result in SignOut.

PinpointTownes commented:
> You can try, but the app identity is unlikely to maintain the hint.

It works.

> Now that I think about it, HandleAuthenticateOnceAsync is unlikely to ever provide a result in SignOut.

Yeah, I realized that too when asking the question :+1: 
When I tested this feature, I was calling `SignOutAsync` from the callback path, where `HandleAuthenticateOnceAsync` still provides a value. I replaced it by a much more realistic test and updated the code in the handler:

```csharp
app.Map("/signout", map => {
    map.Run(async context => {
        var properties = new AuthenticationProperties {
            RedirectUri = "/"
        };

        await context.Authentication.SignOutAsync(OpenIdConnectDefaults.AuthenticationScheme, properties);
    });
});
```

Tratcher commented:
Let's skip the extra option for now, there are other ways it can be dealt with (e.g. in the redirect event, disabling SaveTokensAsClaims, etc.). The hint still won't work with app cookies, but I think only the dev could make that work in the redirect event.
:shipit:

Tratcher commented:
Rebased and merged.

poke commented:
Just as a big FYI to anyone who is using an on-premise AD FS for OpenID Connect, and as a personal reminder (because this was quite painful to figure out…):

You need to **enable `OpenIdConnectOptions.SaveTokens`** if you want the sign-out process to work properly. **Otherwise ADFS will refuse to redirect back** to the configured `SignedOutRedirectUri` and you are stuck on the ADFS logout page.

Apparently, this was changed somewhat recently (probably a few months ago, because it definitely was working for me before). When handling the OAuth logout, ADFS will now verify the logout URL against the configured redirect URIs. In order to do that, it will require the client id. Unfortunately, the only way it can receive the client id is as part of the `id_token_hint`. So we cannot just pass a `client_id` argument but instead have to pass the `id_token_hint` which also requires us to save the token. Exactly that is what this PR is about, so we just have to enable `SaveTokens` in order to allow ADFS to redirect back.
